modules/rgbd/test/test_tsdf.cpp

```diff
@@ -0,0 +1,367 @@
+        {
+            float angle = (float)(CV_2PI * i / framesPerCycle);
+            Affine3f pose;
+            pose = pose.rotate(startPose.rotation());
+            pose = pose.rotate(Vec3f(0.f, -1.f, 0.f) * angle);
+            pose = pose.translate(Vec3f(startPose.translation()[0] * sin(angle),
+                startPose.translation()[1],
+                startPose.translation()[2] * cos(angle)));
+            poses.push_back(pose);
+        }
+
+        return poses;
+    }
+
+};
+
+Ptr<Scene> Scene::create(Size sz, Matx33f _intr, float _depthFactor)
+{
+    return makePtr<SemisphereScene>(sz, _intr, _depthFactor);
+}
+
+static const bool display = false;
+
+void normal_test(bool isHashTSDF, bool isRaycast, bool isFetchPointsNormals, bool isFetchNormals)
+{
+    Ptr<kinfu::Params> _params;
+    if (isHashTSDF)
+        _params = kinfu::Params::hashTSDFParams(true);
+    else
+        _params = kinfu::Params::coarseParams();
+
+    Ptr<Scene> scene = Scene::create(_params->frameSize, _params->intr, _params->depthFactor);
+    std::vector<Affine3f> poses = scene->getPoses();
+
+    Mat depth = scene->depth(poses[0]);
+    UMat _points, _normals, _tmpnormals;
+    UMat _newPoints, _newNormals;
+    Mat  points,  normals;
+    Mat image;
+    AccessFlag af = ACCESS_READ;
+    
+    Ptr<kinfu::Volume> volume = kinfu::makeVolume(_params->volumeType, _params->voxelSize, _params->volumePose, 
+                                _params->raycast_step_factor, _params->tsdf_trunc_dist, _params->tsdf_max_weight, 
+                                _params->truncateThreshold, _params->volumeDims);
+    volume->integrate(depth, _params->depthFactor, poses[0], _params->intr);
+    
+    if (isRaycast)
+    { 
+        volume->raycast(poses[0], _params->intr, _params->frameSize, _points, _normals);
+    }
+    if (isFetchPointsNormals)
+    {
+        volume->fetchPointsNormals(_points, _normals);
+    }
+    if (isFetchNormals)
+    {
+        volume->fetchPointsNormals(_points, _tmpnormals);
+        volume->fetchNormals(_points, _normals);
+    }
+
+    normals = _normals.getMat(af);
+    normals.forEach<Vec4f>([](Vec4f& vector, const int* position)
+        {
+            if (!isnan(vector[0]))
+            {
+                float length = vector[0] * vector[0] +
+                    vector[1] * vector[1] +
+                    vector[2] * vector[2];
+                ASSERT_LT(abs(1 - length), 0.0001f);
+            }
+        });
+    
+    if (display)
+    {
+        imshow("depth", depth * (1.f / _params->depthFactor / 4.f));
+        points = _points.getMat(af);
+        //renderPointsNormals(points, normals, image, _params->lightPose);
+        imshow("render", image);
+        waitKey(30000);
+    }
+    
+    if (isRaycast)
+    {
+    volume->raycast(poses[17], _params->intr, _params->frameSize, _newPoints, _newNormals);
+
+    normals = _newNormals.getMat(af);
+    normals.forEach<Vec4f>([](Vec4f& vector, const int* position)
```

Comment:
The same lambda as above, can be written w/o code duplication

Reply:
done